BPF target support

alessandrod opened the pull request:
This adds `bpfel-unknown-none` and `bpfeb-unknown-none`, two new no_std targets that generate little and big endian BPF. The approach taken is very similar to the cuda target, where `TargetOptions::obj_is_bitcode` is enabled and code generation is done by the linker.

I added the targets to `dist-various-2`. There are [some tests](https://github.com/alessandrod/bpf-linker/tree/main/tests/assembly) in bpf-linker and I'm planning to add more. Those are currently not ran as part of rust CI.

alessandrod committed:
Add BPF target

This change adds the bpfel-unknown-none and bpfeb-unknown-none targets
which can be used to generate little endian and big endian BPF

rust-highfive commented:
r? @matthewjasper

(rust-highfive has picked a reviewer for you, use r? to override)

alessandrod committed:
Fix formatting

matthewjasper commented:
cc @rust-lang/compiler @rust-lang/infra

bjorn3 commented on src/bootstrap/native.rs:
```diff
@@ -154,7 +154,7 @@ impl Step for Llvm {
             None => {
                 "AArch64;ARM;Hexagon;MSP430;Mips;NVPTX;PowerPC;RISCV;\
-                     Sparc;SystemZ;WebAssembly;X86"
+                     Sparc;SystemZ;WebAssembly;X86;BPF"
```
Please update `config.toml.example` too.

alessandrod replied:
Done, thanks!

alessandrod committed:
Add BPF target to config.toml.example

Mark-Simulacrum commented:
I am uncertain that we should be expanding our set of targets right now, I'd really like to get a policy accepted before we do so (cc @joshtriplett). I am inclined to not land this PR at this time.

Mark-Simulacrum commented:
I have asked for feedback on [Zulip](https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/bpf.20targets), but I am still inclined that we not merge these targets at tier 2 status for now. I think I could be sold on tier 3, but someone else would need to approve the target definitions (I don't know much about BPF).

alessandrod commented:
I'm obviously happy for this to be merged at tier 3 (with a little caveat, see below), but here's my rationale for proposing it at tier 2.

The targets meet the criteria explained in https://doc.rust-lang.org/nightly/rustc/platform-support.html#tier-2. `Target`/`TargetOptions` have seen some light refactoring recently but other than that the PR doesn't add anything that would likely cause build breakage at any point. The underlying LLVM BPF target is quite small so it doesn't contribute significantly to build times. It's stable and has been enabled in clang for a long time.

The "emit bitcode, codegen in the linker" approach is identical to the cuda target, which is tier 2. There are a couple of reasons for emitting bitcode. Like NVPTX, BPF supports a subset of Rust (eg, arguments can only be passed in registers, up to a maximum of 5), so it makes sense to push codegen after linking has happened to avoid potentially hitting codegen failures on unused code. Also when targeting older kernels, bpf-linker needs to run more aggressive inlining/loop unrolling optimizations. 

The PR doesn't add new tests. I didn't integrate bpf-linker tests in the rust CI because reviewing the NVPTX PRs I saw that people had some concerns about those, and the NVPTX tests were ultimately disabled. bpf-linker tests are `compiletest` based tho so they can be moved over easily.

The only reason for wanting this at tier 2 sooner rather than later is that while bpf-linker is already usable with rust stable today with `-C linker-flavor=wasm-ld -C linker=bpf-linker`, deployment is a bit of a nightmare. For things to work rustc and bpf-linker need to use the same LLVM. Rust has been tracking the latest LLVM version for a while now, but that's almost never available in linux distributions. With this in tier 2 I could use rustc's libLLVM from bpf-linker, which would simplify things a lot. 

All that said, I wasn't aware of https://github.com/rust-lang/rfcs/pull/2803 when I created the PR, and again I'm happy to wait until that lands.

joshtriplett commented:
@Mark-Simulacrum Working on that policy now.

apiraino commented:
Confirming the decision [during the meeting](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.20meeting.5D.202020-12-31.20.2354818/near/221296802) to wait for the policy to be merged before proceeding.

bors commented:
:umbrella: The latest upstream changes (presumably #80928) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

alessandrod committed:
Pass target features to bpf-linker

alessandrod committed:
Add support for BPF inline assembly

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2389030457) [(plain)](https://github.com/rust-lang/rust/commit/ccae40edff6235ce4777c8c615cdcbc48f3b588d/checks/2389030457/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling rustc_parse_format v0.0.0 (/checkout/compiler/rustc_parse_format)
error[E0583]: file not found for module `bpf`
   --> compiler/rustc_target/src/asm/mod.rs:154:1
    |
154 | mod bpf;
    |
    |
    = help: to create the module `bpf`, create file "compiler/rustc_target/src/asm/bpf.rs"

error[E0432]: unresolved imports `bpf::BpfInlineAsmReg`, `bpf::BpfInlineAsmRegClass`
   --> compiler/rustc_target/src/asm/mod.rs:165:15
    |
165 | pub use bpf::{BpfInlineAsmReg, BpfInlineAsmRegClass};
    |               ^^^^^^^^^^^^^^^  ^^^^^^^^^^^^^^^^^^^^ no `BpfInlineAsmRegClass` in `asm::bpf`
    |               |
    |               no `BpfInlineAsmReg` in `asm::bpf`

error[E0425]: cannot find function `regclass_map` in module `bpf`
   --> compiler/rustc_target/src/asm/mod.rs:677:32
    |
677 |             let mut map = bpf::regclass_map();
    |                                ^^^^^^^^^^^^ not found in `bpf`
help: consider importing one of these items
    |
1   | use crate::asm::aarch64::regclass_map;
    |
---
1   | use crate::asm::mips::regclass_map;
    |
      and 5 other candidates

error[E0425]: cannot find function `fill_reg_map` in module `bpf`
   --> compiler/rustc_target/src/asm/mod.rs:678:18
    |
678 |             bpf::fill_reg_map(arch, has_feature, target, &mut map);
    |                  ^^^^^^^^^^^^ not found in `bpf`
help: consider importing one of these items
    |
1   | use crate::asm::aarch64::fill_reg_map;
    |
    |
1   | use crate::asm::arm::fill_reg_map;
1   | use crate::asm::hexagon::fill_reg_map;
    |
1   | use crate::asm::mips::fill_reg_map;
    |
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #84786) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

dvc94ch commented:
So now that https://github.com/rust-lang/rfcs/pull/2803 was merged, what does that mean for this target?

alessandrod commented:
> So now that [rust-lang/rfcs#2803](https://github.com/rust-lang/rfcs/pull/2803) was merged, what does that mean for this target?

Working on it! A couple of weeks ago I did a rebase and added inline assembly support. As soon as https://reviews.llvm.org/D102118 gets merged I'll do one more rebase and try to get the conversation going.

joshtriplett commented:
I'd be happy to help review any drafts of BPF's answers to the target tier requirements. (I'd suggest full-quoting the tier 3 requirements and responding to each one, which will make it easy to check them off and confirm that BPF is meeting them.)

alessandrod committed:
Add BPF target to platform-support.md

alessandrod commented:
> - A tier 3 target must have a designated developer or developers (the "target
>   maintainers") on record to be CCed when issues arise regarding the target.
>   (The mechanism to track and CC such developers may evolve over time.)

Guess that would be me 🙋‍♂️

> - Targets must use naming consistent with any existing targets; for instance, a
>   target for the same CPU or OS as an existing Rust target should use the same
>   name for that CPU or OS. Targets should normally use the same names and
>   naming conventions as used elsewhere in the broader ecosystem beyond Rust
>   (such as in other toolchains), unless they have a very good reason to
>   diverge. Changing the name of a target can be highly disruptive, especially
>   once the target reaches a higher tier, so getting the name right is important
>   even for a tier 3 target.
>
>   - Target names should not introduce undue confusion or ambiguity unless
>     absolutely necessary to maintain ecosystem compatibility. For example, if
>     the name of the target makes people extremely likely to form incorrect
>     beliefs about what it targets, the name should be changed or augmented to
>     disambiguate it.

The PR introduces two targets: `bpfel-unknown-none` and `bpfeb-unknown-none`. The targets produce [eBPF](https://ebpf.io) code. While BPF and eBPF are technically not the same thing, in the eBPF community the two terms are pretty much used interchangeably.

The LLVM target is called BPF, and the triples used are `bpf`, `bpfel` and `bpfeb`, see the [LLVM target definition](https://github.com/rust-lang/llvm-project/blob/0dbcb3639451a7c20e2d5133b459552281e64455/llvm/lib/Target/BPF/TargetInfo/BPFTargetInfo.cpp#L27). For consistency with LLVM and clang, I opted for `bpfel-` and `bpfeb-`.

As for `unknown-none`, while eBPF is mainly used in Linux, it is not Linux specific. The bytecode produced can be loaded by the [userspace eBPF VM](https://github.com/iovisor/ubpf) and also from the recent [eBPF windows port](https://github.com/microsoft/ebpf-for-windows) from Microsoft.

> - Tier 3 targets may have unusual requirements to build or use, but must not
>   create legal issues or impose onerous legal terms for the Rust project or for
>   Rust developers or users.
>   - The target must not introduce license incompatibilities.

✅

>   - Anything added to the Rust repository must be under the standard Rust
>     license (`MIT OR Apache-2.0`).

✅

>   - The target must not cause the Rust tools or libraries built for any other
>     host (even when supporting cross-compilation to the target) to depend
>     on any new dependency less permissive than the Rust licensing policy. This
>     applies whether the dependency is a Rust crate that would require adding
>     new license exceptions (as specified by the `tidy` tool in the
>     rust-lang/rust repository), or whether the dependency is a native library
>     or binary. In other words, the introduction of the target must not cause a
>     user installing or running a version of Rust or the Rust tools to be
>     subject to any new license requirements.

✅ - the target only requires a custom linker, which is discussed below.

>   - If the target supports building host tools (such as `rustc` or `cargo`),
>     those host tools must not depend on proprietary (non-FOSS) libraries, other
>     than ordinary runtime libraries supplied by the platform and commonly used
>     by other binaries built for the target. For instance, `rustc` built for the
>     target may depend on a common proprietary C runtime library or console
>     output library, but must not depend on a proprietary code generation
>     library or code optimization library. Rust's license permits such
>     combinations, but the Rust project has no interest in maintaining such
>     combinations within the scope of Rust itself, even at tier 3.

Building host tools is not supported.

>   - Targets should not require proprietary (non-FOSS) components to link a
>     functional binary or library.
>   - "onerous" here is an intentionally subjective term. At a minimum, "onerous"
>     legal/licensing terms include but are *not* limited to: non-disclosure
>     requirements, non-compete requirements, contributor license agreements
>     (CLAs) or equivalent, "non-commercial"/"research-only"/etc terms,
>     requirements conditional on the employer or employment of any particular
>     Rust developers, revocable terms, any requirements that create liability
>     for the Rust project or its developers or users, or any requirements that
>     adversely affect the livelihood or prospects of the Rust project or its
>     developers or users.

The targets require [bpf-linker](https://github.com/alessandrod/bpf-linker) which is licensed under MIT OR Apache-2.0.

> - Neither this policy nor any decisions made regarding targets shall create any
>   binding agreement or estoppel by any party. If any member of an approving
>   Rust team serves as one of the maintainers of a target, or has any legal or
>   employment requirement (explicit or implicit) that might affect their
>   decisions regarding a target, they must recuse themselves from any approval
>   decisions regarding the target's tier status, though they may otherwise
>   participate in discussions.
>   - This requirement does not prevent part or all of this policy from being
>     cited in an explicit contract or work agreement (e.g. to implement or
>     maintain support for a target). This requirement exists to ensure that a
>     developer or team responsible for reviewing and approving a target does not
>     face any legal threats or obligations that would prevent them from freely
>     exercising their judgment in such approval, even if such judgment involves
>     subjective matters or goes beyond the letter of these requirements.

✅

> - Tier 3 targets should attempt to implement as much of the standard libraries
>   as possible and appropriate (`core` for most targets, `alloc` for targets
>   that can support dynamic memory allocation, `std` for targets with an
>   operating system or equivalent layer of system-provided functionality), but
>   may leave some code unimplemented (either unavailable or stubbed out as
>   appropriate), whether because the target makes it impossible to implement or
>   challenging to implement. The authors of pull requests are not obligated to
>   avoid calling any portions of the standard library on the basis of a tier 3
>   target not implementing those portions.

The eBPF VM is very restrictive by design. eBPF programs are not general purpose programs, they can't allocate, can't sleep, can't loop indefinitely, etc. Programs generally use only [kernel provided APIs](https://man7.org/linux/man-pages/man7/bpf-helpers.7.html) to do their work and can't really do much more.

That said the parts of `core` that make sense in an eBPF context do work, such as Option, Result core::mem core::ptr etc.

> - The target must provide documentation for the Rust community explaining how
>   to build for the target, using cross-compilation if possible. If the target
>   supports running tests (even if they do not pass), the documentation must
>   explain how to run tests for the target, using emulation if possible or
>   dedicated hardware if necessary.

Cross compilation works with the regular --target={bpfel|bpfeb}-unknown-none. The target does not support running tests.

> - Tier 3 targets must not impose burden on the authors of pull requests, or
>   other developers in the community, to maintain the target. In particular,
>   do not post comments (automated or manual) on a PR that derail or suggest a
>   block on the PR based on a tier 3 target. Do not send automated messages or
>   notifications (via any medium, including via `@`) to a PR author or others
>   involved with a PR regarding a tier 3 target, unless they have opted into
>   such messages.
>   - Backlinks such as those generated by the issue/PR tracker when linking to
>     an issue or PR are not considered a violation of this policy, within
>     reason. However, such messages (even on a separate repository) must not
>     generate notifications to anyone involved with a PR who has not requested
>     such notifications.
> - Patches adding or updating tier 3 targets must not break any existing tier 2
>   or tier 1 target, and must not knowingly break another tier 3 target without
>   approval of either the compiler team or the maintainers of the other tier 3
>   target.
>   - In particular, this may come up when working on closely related targets,
>     such as variations of the same architecture with different features. Avoid
>     introducing unconditional uses of features that another variation of the
>     target may not have; use conditional compilation or runtime detection, as
>     appropriate, to let each target run code supported by that target.

Sounds reasonable ✅

> If a tier 3 target stops meeting these requirements, or the target maintainers
> no longer have interest or time, or the target shows no signs of activity and
> has not built for some time, or removing the target would improve the quality
> of the Rust codebase, we may post a PR to remove it; any such PR will be CCed
> to the target maintainers (and potentially other people who have previously
> worked on the target), to check potential interest in improving the situation.

Also reasonable ✅

dvc94ch commented:
I thought that tier2 status is required for the bpf-linker to be able to share the rust llvm backend. Has this changed in the meantime?

alessandrod commented:
> I thought that tier2 status is required for the bpf-linker to be able to share the rust llvm backend. Has this changed in the meantime?

Yes but quoting the RFC:

> Note that a target must have already received approval for the next lower tier, and spent a reasonable amount of time at that tier, before making a proposal for promotion to the next higher tier; this is true even if a target meets the requirements for several tiers at once. This policy leaves the precise interpretation of "reasonable amount of time" up to the approving teams; those teams may scale the amount of time required based on their confidence in the target and its demonstrated track record at its current tier. At a minimum, multiple stable releases of Rust should typically occur between promotions of a target.

Spending a few release cycles at tier 3 seems reasonable to me.

bors commented:
:umbrella: The latest upstream changes (presumably #85594) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

alessandrod committed:
Update LLVM submodule

alessandrod commented:
Btw I just realized that this is still marked as s-blocked but that can be removed now

joshtriplett commented:
Does any other compiler (e.g. clang) already use the triples `bpfel-unknown-none` or `bpfeb-unknown-none`? If not, could we change those to just `bpfel-none` and `bpfeb-none`? The middle component of the triple is a mostly historical curiosity that doesn't tend to apply to any current target.

alessandrod commented:
> Does any other compiler (e.g. clang) already use the triples `bpfel-unknown-none` or `bpfeb-unknown-none`? If not, could we change those to just `bpfel-none` and `bpfeb-none`? The middle component of the triple is a mostly historical curiosity that doesn't tend to apply to any current target.

No clang/LLVM don't use bpf*-unknown-none, so happy to rename. Apparently tho gcc uses/will use `bpf-unknown-none` https://lwn.net/Articles/800606/.

nagisa commented on compiler/rustc_target/src/abi/call/bpf.rs:
```diff
@@ -0,0 +1,31 @@
+// see BPFCallingConv.td
```
This is not all that useful a reference to a typical Rust developer who doesn't dabble in LLVM. Could this perhaps be replaced with a link to this file?

alessandrod replied:
✅

nagisa commented on compiler/rustc_target/src/abi/call/bpf.rs:
```diff
@@ -0,0 +1,31 @@
+    if ret.layout.is_aggregate() || ret.layout.size.bits() > 64 {
+        ret.make_indirect();
+    } else {
+        ret.extend_integer_width_to(64);
```
AFAIU when alu32 feature is enabled, the 32-bit values are expected to be passed in the `w*` registers, which I think means that we want to conditionally extend to 32 bits here when that feature is enabled.

Perhaps you could always extend to just 32 bits and let LLVM handle extending to 64 bits if alu32 is not enabled?

alessandrod replied:
Yes good point, I've done that.

nagisa reviewed (commented):
* Please add tests for assembly support (see `src/test/assembly/asm` for examples);
* If possible a couple codegen tests for basic operation (e.g. you mentioned that some of the core works);
* A test for alu32 feature would be nice (I think in writing such a test you may hit an oversight in that `compiler/rustc_typeck/src/collect.rs` was not adjusted for the newly added target features.

alessandrod committed:
BPF: add assembly test

alessandrod committed:
BPF: fix #[target_feature(enable = "alu32")]

alessandrod committed:
BPF: abi: extend args/ret to 32 bits

Let LLVM extend to 64 bits when alu32 is not enabled

alessandrod committed:
BPF: misc minor review fixes

alessandrod commented on src/tools/compiletest/src/util.rs:
```diff
@@ -48,6 +48,8 @@ const ARCH_TABLE: &[(&str, &str)] = &[
     ("asmjs", "asmjs"),
     ("avr", "avr"),
+    ("bpfeb", "bpfeb"),
+    ("bpfel", "bpfel"),
```
✅

alessandrod commented on src/tools/compiletest/src/runtest.rs:
```diff
@@ -2310,6 +2311,10 @@ impl<'test> TestCx<'test> {
                 // No extra flags needed.
             }
 
+            Some("bpf-linker") => {
```
✅

alessandrod commented on src/doc/unstable-book/src/library-features/asm.md:
```diff
@@ -615,6 +618,8 @@ Each register class has constraints on which value types they can be used with.
 | PowerPC | `freg` | None | `f32`, `f64` |
 | wasm32 | `local` | None | `i8` `i16` `i32` `i64` `f32` `f64` |
+| BPF | `reg` | None | `i8` `i16` `i32` `i64` |
+| BPF | `wreg` | `alu32` | `i8` `i16` `i32`|
```
✅

alessandrod commented on src/doc/unstable-book/src/library-features/asm.md:
```diff
@@ -570,6 +571,8 @@ Here is the list of currently supported register classes:
 | PowerPC | `freg` | `f[0-31]` | `f` |
 | wasm32 | `local` | None\* | `r` |
+| BPF | `reg` | `r[0-10]` | `r`|
+| BPF | `wreg` | `w[0-10]` | `w`|
```
✅

alessandrod commented on src/bootstrap/native.rs:
```diff
@@ -236,7 +236,7 @@ impl Step for Llvm {
             None => {
                 "AArch64;ARM;Hexagon;MSP430;Mips;NVPTX;PowerPC;RISCV;\
-                     Sparc;SystemZ;WebAssembly;X86"
+                     Sparc;SystemZ;WebAssembly;X86;BPF"
```
✅

alessandrod commented on compiler/rustc_target/src/spec/bpf_base.rs:
```diff
@@ -0,0 +1,38 @@
+        no_builtins: true,
+        panic_strategy: PanicStrategy::Abort,
+        position_independent_executables: true,
+        merge_functions: MergeFunctions::Disabled,
```
✅

alessandrod commented:
Thanks for the review!

>     * Please add tests for assembly support (see `src/test/assembly/asm` for examples);

Done
 
>     * A test for alu32 feature would be nice (I think in writing such a test you may hit an oversight in that `compiler/rustc_typeck/src/collect.rs` was not adjusted for the newly added target features.

Yup nice catch, I fixed `compiler/rustc_typeck/src/collect.rs` and added a test.

>     * If possible a couple codegen tests for basic operation (e.g. you mentioned that some of the core works);

I'll do this next, trying to come up with something sensible to test, ideas welcome :)

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2700700362) [(plain)](https://github.com/rust-lang/rust/commit/ab93a139ef110f977f08d5bedc744654efe729f6/checks/2700700362/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
* highest error code: E0783
Found 517 error codes
Found 0 error codes with no tests
Done!
tidy error: Found 1 features without a gate test.
Expected a gate test for the feature 'bpf_target_feature'.
Hint: create a failing test file named 'feature-gate-bpf_target_feature.rs'
      in the 'ui' test suite, with its failures due to
      missing usage of `#![feature(bpf_target_feature)]`.
Hint: If you already have such a test and don't want to rename it,
      you can also add a // gate-test-bpf_target_feature line to the test file.
some tidy checks failed


command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/rust-tidy" "/checkout" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "/checkout/obj/build" "16"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test --stage 2 src/tools/tidy
Build completed unsuccessfully in 0:00:13
```

</details>

alessandrod committed:
Fix tests

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2700795290) [(plain)](https://github.com/rust-lang/rust/commit/5593d8d0f67b0a384a201a6e39c33429f2b36cd5/checks/2700795290/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

---- [ui] ui/target-feature/gate.rs stdout ----
diff of stderr:

1 error[E0658]: the target feature `avx512bw` is currently unstable
-   --> $DIR/gate.rs:31:18
+   --> $DIR/gate.rs:32:18
3    |
4 LL | #[target_feature(enable = "avx512bw")]


The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/target-feature/gate/gate.stderr
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/target-feature/gate/gate.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args target-feature/gate.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/target-feature/gate.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/target-feature/gate" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/target-feature/gate/auxiliary"
Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
error[E0658]: the target feature `avx512bw` is currently unstable
   |
   |
LL | #[target_feature(enable = "avx512bw")]
   |
   = note: see issue #44839 <https://github.com/rust-lang/rust/issues/44839> for more information
   = note: see issue #44839 <https://github.com/rust-lang/rust/issues/44839> for more information
   = help: add `#![feature(avx512_target_feature)]` to the crate attributes to enable
error: aborting due to previous error

For more information about this error, try `rustc --explain E0658`.

---
test result: FAILED. 11836 passed; 1 failed; 97 ignored; 0 measured; 0 filtered out; finished in 122.69s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/ui" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "ui" "--mode" "ui" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-10/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "10.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine frontendopenmp fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcerror orcjit passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvutils runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86utils xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap --stage 2 test --exclude src/tools/tidy
Build completed unsuccessfully in 0:12:38
```

</details>

bjorn3 commented on compiler/rustc_feature/src/active.rs:
```diff
@@ -250,6 +250,7 @@ declare_features! (
     (active, f16c_target_feature, "1.36.0", Some(44839), None),
     (active, riscv_target_feature, "1.45.0", Some(44839), None),
     (active, ermsb_target_feature, "1.49.0", Some(44839), None),
+    (active, bpf_target_feature, "1.53.0", Some(44839), None),
```
```suggestion
    (active, bpf_target_feature, "1.54.0", Some(44839), None),
```

bjorn3 commented on config.toml.example:
```diff
@@ -94,7 +94,7 @@ changelog-seen = 2
 # likely, teach rustc about the C ABI of the target. Get in touch with the
 # Rust team and file an issue if you need assistance in porting!
-#targets = "AArch64;ARM;Hexagon;MSP430;Mips;NVPTX;PowerPC;RISCV;Sparc;SystemZ;WebAssembly;X86"
+#targets = "AArch64;ARM;Hexagon;MSP430;Mips;NVPTX;PowerPC;RISCV;Sparc;SystemZ;WebAssembly;X86;BPF"
```
```suggestion
#targets = "AArch64;ARM;BPF;Hexagon;MSP430;Mips;NVPTX;PowerPC;RISCV;Sparc;SystemZ;WebAssembly;X86"
```

bjorn3 commented on src/test/codegen/bpf-alu32.rs:
```diff
@@ -0,0 +1,12 @@
+// compile-flags: --emit=asm --target bpfel-unknown-none
```
Shouldn't this be `--emit=llvm-ir`?

alessandrod replied:
Oops yes that's right, I started the test as an asm test and then moved it over. I've removed the whole line since it isn't really needed anymore.

bjorn3 commented:
What will the crate type for the actual program be? Will it be `cdylib`? But the linker code doesn't handle `cdylib`. It can't be `rlib` as that has an unstable implementation. It can't be `staticlib` as that doesn't get linked. And I don't think it can be `bin` as that only exports a single `main` function that should take an `argc` and `argv`.

alessandrod committed:
BPF: remove unnecessary flags from src/test/codegen/bpf-alu32.rs

alessandrod committed:
BPF: review fixes

alessandrod commented:
> What will the crate type for the actual program be? Will it be `cdylib`? But the linker code doesn't handle `cdylib`. It can't be `rlib` as that has an unstable implementation. It can't be `staticlib` as that doesn't get linked. And I don't think it can be `bin` as that only exports a single `main` function that should take an `argc` and `argv`.

The linker can't link in external `cdylib`s but can output one, and `#[no_main]` bins work too, as `#[no_mangle]` items get exported. In both cases the output is actually a static ELF object.

nagisa commented:
@bors r+

bors commented:
:pushpin: Commit 0adb9331ffab305feeb07fb57b1f189e35cb5adf has been approved by `nagisa`

<!-- @bors r=nagisa 0adb9331ffab305feeb07fb57b1f189e35cb5adf -->
<!-- homu: {"type":"Approved","sha":"0adb9331ffab305feeb07fb57b1f189e35cb5adf","approver":"nagisa"} -->

bors commented:
:hourglass: Testing commit 0adb9331ffab305feeb07fb57b1f189e35cb5adf with merge f434217aab9abf583ebc928b97ab4116921137aa...
<!-- homu: {"type":"BuildStarted","head_sha":"0adb9331ffab305feeb07fb57b1f189e35cb5adf","merge_sha":"f434217aab9abf583ebc928b97ab4116921137aa"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/2755570855)
Approved by: nagisa
Pushing f434217aab9abf583ebc928b97ab4116921137aa to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"nagisa","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/2755570855"},"merge_sha":"f434217aab9abf583ebc928b97ab4116921137aa"} -->

alessandrod commented:
Thanks everyone who helped with this 🙏

dvc94ch commented:
Thank you for persevering for 6 months.

danobi commented:
Sorry if this is the wrong place to ask (please redirect me if appropriate).

I'd like to better understand the rationale behind having rustc emit bitcode and the linker doing codegen. There's already prior art in the bpf ecosystem to link ELF objects together ( https://github.com/torvalds/linux/blob/de5540965853e514a85d3b775e9049deb85a2ff3/tools/bpf/bpftool/Documentation/bpftool-gen.rst?plain=1#L28-L31 ).

Admittedly, I don't think it's much of an issue if bpf-linker gets merged into upstream rust.

> There are a couple of reasons for emitting bitcode. Like NVPTX, BPF supports a subset of Rust (eg, arguments can only be passed in registers, up to a maximum of 5), so it makes sense to push codegen after linking has happened to avoid potentially hitting codegen failures on unused code. 

Could you clarify this point? I'm ignorant as to how this would be different w.r.t. clang and bpf progs written in C.

> Also when targeting older kernels, bpf-linker needs to run more aggressive inlining/loop unrolling optimizations.

Could you also clarify this too? I'm not sure how separating out codegen into another binary helps with this issue. It seems more reliable to solve this issue by letter the user annotate their code (always_unroll, never_unroll, etc).

bjorn3 commented:
> Could you clarify this point? I'm ignorant as to how this would be different w.r.t. clang and bpf progs written in C.

C can be used without a standard library that may use unsupported constructs. Rust requires at least libcore to be compiled. Without it, even something as simple as defining a variable complains about a missing `Sized` lang item. Or adding two numbers complains about a eg missing `impl Add for u8` in case of `u8`s.

> Could you also clarify this too? I'm not sure how separating out codegen into another binary helps with this issue. It seems more reliable to solve this issue by letter the user annotate their code (always_unroll, never_unroll, etc).

Libcore is pre-compiled for a target independent of which kernel is used. This pre-compiled code may not have unrolled loops/inlined functions enough for the running kernel. Moving the codegen into the linker allows the linker to codegen eBPF code accepted by the specific kernel it is going to run on.

danobi commented:
> > Could you clarify this point? I'm ignorant as to how this would be different w.r.t. clang and bpf progs written in C.
> 
> C can be used without a standard library that may use unsupported constructs. Rust requires at least libcore to be compiled. Without it, even something as simple as defining a variable complains about a missing `Sized` lang item. Or adding two numbers complains about a eg missing `impl Add for u8` in case of `u8`s.

Thanks for the explanation. That makes sense to me.

> > Could you also clarify this too? I'm not sure how separating out codegen into another binary helps with this issue. It seems more reliable to solve this issue by letter the user annotate their code (always_unroll, never_unroll, etc).
> 
> Libcore is pre-compiled for a target independent of which kernel is used. This pre-compiled code may not have unrolled loops/inlined functions enough for the running kernel. Moving the codegen into the linker allows the linker to codegen eBPF code accepted by the specific kernel it is going to run on.

So is the idea w/ bpf-linker in a separate binary (rather than embedded in rustc) to let bpf-linker move faster and support more kernels than at rustc release pace? Alternatively, would it ever make sense to embed bpf-linker inside rustc?

bjorn3 commented:
> So is the idea w/ bpf-linker in a separate binary (rather than embedded in rustc) to let bpf-linker move faster and support more kernels than at rustc release pace?

Not necessarily. The main point as far as I understand is pushing all the codegen to the compilation of the bpf program when the user can choose which kernel to target instead of having part of the codegen happen on rust's CI where it is not known which kernel will be targeted by the user.

> Alternatively, would it ever make sense to embed bpf-linker inside rustc?

Maybe?
